src/storage/sqlite.rs

```diff
@@ -91,24 +93,59 @@ impl SqliteStorage {
             let id: i32 = row.get(0).map_err(bad_storage)?;
             let atype: CK_ULONG = row.get(1).map_err(bad_storage)?;
-            let value = row
-                .get_ref(2)
-                .map_err(bad_storage)?
-                .as_blob()
-                .map_err(bad_code)?;
+            let val = row.get_ref(2).map_err(bad_storage)?;
             /* TODO: enc */
             if objid != id {
                 objid = id;
                 objects.push(Object::new());
             }
             if let Some(obj) = objects.last_mut() {
-                let ck_attr = CK_ATTRIBUTE {
-                    type_: atype,
-                    pValue: value.as_ptr() as *mut _,
-                    ulValueLen: value.len() as CK_ULONG,
+                let attrtype = attribute::attr_id_to_attrtype(atype)?;
+                let attr = match attrtype {
+                    AttrType::BoolType => match val
+                        .as_i64_or_null()
+                        .map_err(bad_storage)?
+                    {
+                        Some(b) => attribute::from_bool(atype, b != 0),
+                        None => return err_rv!(CKR_ATTRIBUTE_VALUE_INVALID),
+                    },
+                    AttrType::NumType => match val
+                        .as_i64_or_null()
+                        .map_err(bad_storage)?
+                    {
+                        Some(n) => attribute::from_ulong(atype, n as CK_ULONG),
+                        None => return err_rv!(CKR_ATTRIBUTE_VALUE_INVALID),
+                    },
+                    AttrType::StringType => match val
+                        .as_str_or_null()
+                        .map_err(bad_storage)?
+                    {
+                        Some(s) => attribute::from_string(atype, s.to_string()),
+                        None => return err_rv!(CKR_ATTRIBUTE_VALUE_INVALID),
+                    },
+                    AttrType::BytesType => {
+                        match val.as_blob_or_null().map_err(bad_storage)? {
+                            Some(v) => attribute::from_bytes(atype, v.to_vec()),
+                            None => attribute::from_bytes(atype, Vec::new()),
+                        }
+                    }
+                    AttrType::DateType => match val
+                        .as_str_or_null()
+                        .map_err(bad_storage)?
+                    {
+                        Some(s) => attribute::from_date(
+                            atype,
+                            attribute::string_to_ck_date(s)?,
+                        ),
+                        None => return err_rv!(CKR_ATTRIBUTE_VALUE_INVALID),
+                    },
+                    AttrType::DenyType => {
+                        return err_rv!(CKR_ATTRIBUTE_TYPE_INVALID)
+                    }
+                    AttrType::IgnoreType => {
+                        return err_rv!(CKR_ATTRIBUTE_TYPE_INVALID)
+                    }
```

Comment:
These two branches have the same content so it might be cleaner to merge them (not sure about the exact syntax now)
```suggestion
                    AttrType::DenyType,
                    AttrType::IgnoreType => {
                        return err_rv!(CKR_ATTRIBUTE_TYPE_INVALID)
                    }
```
Or catch them as `_`